### PR TITLE
GEOMESA-1864 Refactor HBase coprocessor use and expose aggregator API…

### PR DESCRIPTION
**Pull request by jnh5y**
…Squashed commit of the following:

* Create one coprocessor for GeoMesa HBase distributed processing.
* Exposes GeoMesaHBaseAggregator interface.
* Refactors Heatmap computations to use the new interface.

Signed-off-by: Jim Hughes <jnh5y@ccri.com>

### CODE REVIEW

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/coprocessor/GeoMesaCoprocessor.scala [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+import scala.annotation.tailrec
+import scala.collection.JavaConversions._
+
+class GeoMesaCoprocessor extends GeoMesaCoprocessorService with Coprocessor with CoprocessorService {
+
+  private var env: RegionCoprocessorEnvironment = _
+
+  @throws[IOException]
+  def start(env: CoprocessorEnvironment): Unit = {
```
Comment (elahrvivaz) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/coprocessor/GeoMesaCoprocessor.scala]: can you add 'override' markers to the methods that are overridden? makes it a lot easier to reason about things

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/coprocessor/GeoMesaCoprocessor.scala [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+  }
+
+  @throws[IOException]
+  def stop(coprocessorEnvironment: CoprocessorEnvironment): Unit = {
+  }
+
+  def getService: Service = this
+
+  def getResult(controller: RpcController,
+                 request: GeoMesaProto.GeoMesaCoprocessorRequest,
```
Comment (elahrvivaz) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/coprocessor/GeoMesaCoprocessor.scala]: [picky] alignment

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/coprocessor/GeoMesaCoprocessor.scala [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+        scan.setFilter(filterList)
+        // TODO: Explore use of MultiRangeFilter
+        val scanner = env.getRegion.getScanner(scan)
+
+        try {
+          val results = new java.util.ArrayList[Cell]
+
+          def processResults() = {
+            for (cell <- results) {
+              val sf = serializer.deserialize(cell.getValueArray, cell.getValueOffset, cell.getValueLength)
```
Comment (elahrvivaz) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/coprocessor/GeoMesaCoprocessor.scala]: we should add a TODO here to set the feature id

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/coprocessor/GeoMesaCoprocessor.scala [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+              processResults()
+              readScanner()
+            } else {
+              processResults()
+            }
+          }
+
+          readScanner()
+        } catch {
+          case t: Throwable => throw t
```
Comment (elahrvivaz) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/coprocessor/GeoMesaCoprocessor.scala]: can remove this catch clause, and add {} around the finally clause

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/coprocessor/GeoMesaCoprocessor.scala (hunk 3 above)
Comment (elahrvivaz) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/coprocessor/GeoMesaCoprocessor.scala]: are we dealing with transforms here? think we'll need some extra work to account for that. can add a TODO for now

Comment (jnh5y) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/coprocessor/GeoMesaCoprocessor.scala]: Transforms are handled by the JSimpleFeatureFilter.

Comment (elahrvivaz) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/coprocessor/GeoMesaCoprocessor.scala]: is the sft here the correct transform sft then?

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala [hunk 5]
```diff
@@ -73,43 +70,45 @@ case class ScanPlan(filter: HBaseFilterStrategyType,
...
-      // note: mrrf first priority
-      val filterList = new FilterList(remoteFilters.sortBy(_._1).map(_._2).+:(mrrf): _*)
-
-      val scan = new Scan()
-      scan.setFilter(filterList)
-
-      val is: Map[String, String] = KryoLazyDensityCoprocessor.configure(sft, scan, filterList, hints)
-      val byteArray: Array[Byte] = KryoLazyDensityCoprocessor.serializeOptions(is)
+    // TODO: Refactor this logical into HBasePlatform?
+      val (scan, filterList) = calculateScanAndFilterList(ranges, remoteFilters)
```
Comment (elahrvivaz) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala]: [picky] looks like indentation is off

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/coprocessor/GeoMesaCoprocessor.scala (hunk 3 above)
Comment (jnh5y) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/coprocessor/GeoMesaCoprocessor.scala]: Actually.... no.  Lemme take a crack at that.

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/coprocessor/GeoMesaCoprocessor.scala (hunk 1 above)
Comment (jnh5y) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/coprocessor/GeoMesaCoprocessor.scala]: Yes!

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/coprocessor/GeoMesaCoprocessor.scala (hunk 4 above)
Comment (jnh5y) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/coprocessor/GeoMesaCoprocessor.scala]: Of Course

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala (hunk 5 above)
Comment (jnh5y) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala]: Derp; fixing it!

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/coprocessor/aggregators/HBaseDensityAggregator.scala [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+import org.locationtech.geomesa.hbase.coprocessor.SFT_OPT
+import org.locationtech.geomesa.index.utils.KryoLazyDensityUtils
+import org.locationtech.geomesa.index.utils.KryoLazyDensityUtils._
+import org.locationtech.geomesa.utils.geotools.{GeometryUtils, SimpleFeatureTypes}
+import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
+
+import scala.collection.mutable
+
+class HBaseDensityAggregator extends GeoMesaHBaseAggregator {
+  val densityUtils = new KryoLazyDensityUtils {}
```
Comment (elahrvivaz) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/coprocessor/aggregators/HBaseDensityAggregator.scala]: [minor] now that this is a new instance per request, we could just extend KryoLazyDensityUtils

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/coprocessor/GeoMesaCoprocessor.scala [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    * @param table
+    * @return HashMap result;
+    * @throws Throwable
+    */
+  def execute(table: Table, options: Array[Byte]): List[ByteString] = {
+    val requestArg: GeoMesaCoprocessorRequest = GeoMesaCoprocessorRequest.newBuilder().setOptions(ByteString.copyFrom(options)).build()
+
+    val callBack: GeoMesaHBaseCallBack = new GeoMesaHBaseCallBack()
+
+    table.coprocessorService(classOf[GeoMesaCoprocessorService], null, null, new Call[GeoMesaCoprocessorService, ByteString]() {
```
Comment (aheyne) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/coprocessor/GeoMesaCoprocessor.scala]: Is there any reason to not pull the new Call into variable so this is easier to read?

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/coprocessor/GeoMesaCoprocessor.scala [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+        }
+        response.getSf
+      }
+    },
+      callBack
+    )
+    callBack.getResult
+  }
+
+  def getAggregator(options: Map[String, String]): GeoMesaHBaseAggregator = {
```
Comment (aheyne) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/coprocessor/GeoMesaCoprocessor.scala]: Should we move this to coprocessor/package.scala? Similar to ones there.

Comment (jnh5y) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/coprocessor/GeoMesaCoprocessor.scala]: I don't think so.  I'll mark this as package private to indicate scope more clearly.

**Pull request merged**